controllers/registrycredential_controller.go

```diff
@@ -0,0 +1,216 @@
+		if apierrors.IsNotFound(err) {
+			log.Info("creating new registry credential secret")
+			return r.createSecret(ctx, credentials)
+		}
+		return err
+	}
+
+	return r.updateSecret(ctx, log, credentials, existingSecret)
+}
+
+func (r *RegistryCredentialsReconciler) updateSecret(ctx context.Context, log logr.Logger, credentials platformv1alpha1.RegistryCredential, secret *corev1.Secret) error {
+	if secret.Data == nil {
+		return fmt.Errorf("secret data can not be nil")
+	}
+
+	existingAuths := string(secret.Data[dockerconfigjson])
+	expectedAuths, err := r.genSecretAuths(ctx, credentials)
+	if err != nil {
+		return err
+	}
+	if existingAuths != expectedAuths {
+		secret.Data[dockerconfigjson] = []byte(expectedAuths)
+		log.Info("updating registry credential secret")
+		return r.Update(ctx, secret)
+	}
+
+	return nil
+}
+
+func (r *RegistryCredentialsReconciler) createSecret(ctx context.Context, credentials platformv1alpha1.RegistryCredential) error {
+	secret := &corev1.Secret{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      credentials.GetPasswordSecretName(),
+			Namespace: credentials.Namespace,
+			Labels:    map[string]string{secretSyncLabel: "all", secretTypeLabel: "plural-creds"},
+		},
+		Type: "kubernetes.io/dockerconfigjson",
+	}
+
+	auths, err := r.genSecretAuths(ctx, credentials)
+	if err != nil {
+		return err
+	}
+	secret.Data = map[string][]byte{dockerconfigjson: []byte(auths)}
+
+	return r.Create(ctx, secret)
+}
+
+func (r *RegistryCredentialsReconciler) genSecretAuths(ctx context.Context, credentials platformv1alpha1.RegistryCredential) (string, error) {
+	passwordSecret := &corev1.Secret{}
+	if err := r.Get(ctx, client.ObjectKey{Namespace: credentials.Namespace, Name: credentials.Spec.PasswordSecretRef.Name}, passwordSecret); err != nil {
+		return "", err
+	}
+	password := passwordSecret.Data[credentials.Spec.PasswordSecretRef.Key]
+
+	rawAuth := fmt.Sprintf("%s:%s", credentials.Spec.Username, password)
+	auth := base64.StdEncoding.EncodeToString([]byte(rawAuth))
+
+	a := Auths{
+		Cred: map[string]RegistryCred{credentials.Spec.Server: {
+			Auth:     auth,
+			Password: string(password),
+			Username: credentials.Spec.Username,
+			Email:    credentials.Spec.Email,
+		}},
+	}
+
+	rawAuths, err := json.Marshal(a)
+	if err != nil {
+		return "", err
+	}
+	auths := base64.StdEncoding.EncodeToString(rawAuths)
+	return auths, nil
+}
+
+func (r *RegistryCredentialsReconciler) deleteRegistryCredentialSecret(ctx context.Context, credentials platformv1alpha1.RegistryCredential) error {
+	secret := &corev1.Secret{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      credentials.GetPasswordSecretName(),
+			Namespace: credentials.Namespace,
+		},
+	}
+	return r.Delete(ctx, secret)
+}
+
+// requestFromSecret returns a reconcile.Request for the credential registry if the secret is a password reference.
+func requestFromSecret(c client.Client) handler.EventHandler {
```

Comment:
can't we synthesize this with an owner reference or does that not necessarily work?  I'm fine with this approach, just curious.

Reply:
the owner reference will work during deletion not necessarily with an update.

Reply:
gotcha good to know